cache/models/samples.js

```diff
@@ -368,70 +389,54 @@ function upsertOneSample(sampleQueryBodyObj, method) {
-    'hgetall',
-    sampleStore.toKey(constants.objectType.sample, sampleName),
-  ]);
-
   let aspectObj;
-  return new Promise((resolve) => {
-    redisClient.sismemberAsync(constants.indexKey.subject, subjKey)
-    .then((isSubjPresent) => {
-      if (!isSubjPresent) {
-        const err = new redisErrors.ResourceNotFoundError({
-          explanation: 'Subject not found.',
-        });
-        errObj.explanation = err;
-        resolve(errObj);
-      }
 
-      return redisClient.hgetallAsync(
-        sampleStore.toKey(constants.objectType.aspect, aspectName)
-      );
-    })
-    .then((aspect) => {
-      if (!aspect) {
-        const err = new redisErrors.ResourceNotFoundError({
-          explanation: 'Aspect not found.',
-        });
-        errObj.explanation = err;
-        resolve(errObj);
-      }
+  // if any of the promise errors, the subsequent promise does not process and
+  // error is returned, else sample is returned
+  return Promise.all([
+    redisClient.sismemberAsync(constants.indexKey.subject, subjKey),
+    redisClient.hgetallAsync(
+      sampleStore.toKey(constants.objectType.aspect, aspectName)
+    ),
+    redisClient.hgetallAsync(sampleKey),
+  ])
+  .then((responses) => {
+    const [isSubjPresent, aspect, sample] = responses;
+    if (!isSubjPresent) {
+      handleUpsertError(constants.objectType.subject, isBulk);
+    }
 
-      aspectObj = sampleStore.arrayStringsToJson(
-        aspect, constants.fieldsToStringify.aspect
-      );
+    if (!aspect) {
+      handleUpsertError(constants.objectType.aspect, isBulk);
+    }
 
-      return redisClient.hgetallAsync(
-        sampleStore.toKey(constants.objectType.sample, sampleName)
-      );
-    })
-    .then((sample) => {
-      // sampleQueryBodyObj updated with fields
-      createSampHsetCommand(sampleQueryBodyObj, sample, aspectObj);
-
-      // add aspect name to subject set, add sample key to sample set,
-      // create/update hash of sample
-      return redisClient.batch([
-        ['sadd', subjKey, aspectName],
-        ['sadd', constants.indexKey.sample, sampleKey],
-        ['hmset', sampleKey, sampleQueryBodyObj],
-      ]).execAsync();
-    })
-    .then(() => redisClient.hgetallAsync(
-        sampleStore.toKey(constants.objectType.sample, sampleName)
-    ))
-    .then((updatedSamp) => {
-      // attach aspect to sample
-      const resSampAsp = cleanAddAspectToSample(updatedSamp, aspectObj, method);
-      resolve(resSampAsp);
-    });
+    aspectObj = sampleStore.arrayStringsToJson(
+      aspect, constants.fieldsToStringify.aspect
+    );
+
+    // sampleQueryBodyObj updated with fields
+    createSampHsetCommand(sampleQueryBodyObj, sample, aspectObj);
+
+    // add aspect name to subject set, add sample key to sample set,
+    // create/update hash of sample
+    return redisClient.batch([
+      ['sadd', subjKey, aspectName],
+      ['sadd', constants.indexKey.sample, sampleKey],
+      ['hmset', sampleKey, sampleQueryBodyObj],
```

Comment:
Just to optimize, I think we should execute only hmset if sample is already present, let me make that change as well.